sklearn/metrics/pairwise.py

```diff
@@ -585,6 +592,119 @@ def cosine_distances(X, Y=None):
     return S
 
 
+def gower_distances(X, Y=None, w=None, categorical_features=None):
+    """
+    Computes the gower distances between X and Y
+
+    Read more in the :ref:`User Guide <metrics>`.
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+
+    Y : array-like, shape (n_samples, n_features)
+
+    w:  array-like, shape (n_features)
+    According the Gower formula, w is an attribute weight.
+
+    categorical_features: array-like, shape (n_features)
+    Indicates with True/False wheter a column is a categorical attribute.
+    This is useful when categorical atributes are represented as integer
+    values.
+
+    Returns
+    -------
+    similarities : ndarray, shape (n_samples, )
+
+    Notes
+    ------
+    Gower is a similarity measure for categorical, boolean and numerical mixed
+    data.
+
```

Comment:
Notes should state that non-numeric features, and numeric feature ranges are determined from X and not Y. If that's the logic we think is sensible.

Reply:
I feel uncomfortable about having ranges automatically determined because it means that the metric depends on how we split the dataset; it treats one portion as "training data". I would rather either (a) the user be forced to do any scaling beforehand; or (b) we allow (perhaps require) the user to specify the range parameters...

Reply:
I wouldn't be opposed to expecting the user to scale the input values beforehand, and raising a `ValueError` if that's not the case. The user can easily put a scaler before calling this function in a pipeline anyway. That way we keep the implementation easy, and keep the _transformer_ logic out of the distance function.